src/Course/StateT.hs

```diff
@@ -60,102 +52,60 @@ instance Functor f => Functor (StateT s f) where
 -- >>> runStateT (StateT (\s -> ((+2), s P.++ [1]) :. ((+3), s P.++ [1]) :. Nil) <*> (StateT (\s -> (2, s P.++ [2]) :. Nil))) [0]
 -- [(4,[0,1,2]),(5,[0,1,2])]
 instance Monad f => Applicative (StateT s f) where
-  pure ::
-    a
-    -> StateT s f a
-  pure =
-    error "todo: Course.StateT pure#instance (StateT s f)"
-  (<*>) ::
-   StateT s f (a -> b)
-    -> StateT s f a
-    -> StateT s f b
-  (<*>) =
-    error "todo: Course.StateT (<*>)#instance (StateT s f)"
-
--- | Implement the `Monad` instance for @StateT s f@ given a @Monad f@.
--- Make sure the state value is passed through in `bind`.
---
--- >>> runStateT ((const $ putT 2) =<< putT 1) 0
--- ((),2)
---
+  pure :: a -> StateT s f a
+  pure a = StateT (\s -> pure (a, s))
+  (StateT stateF) <*> (StateT stateV) = StateT (\s -> stateV s >>= doApp)
+    where
+      doApp (a, s) = stateF s >>= (\(f', s') -> pure (f' a, s'))
+
 -- >>> let modify f = StateT (\s -> pure ((), f s)) in runStateT (modify (+1) >>= \() -> modify (*2)) 7
 -- ((),16)
 instance Monad f => Monad (StateT s f) where
-  (=<<) ::
-    (a -> StateT s f b)
-    -> StateT s f a
-    -> StateT s f b
-  (=<<) =
-    error "todo: Course.StateT (=<<)#instance (StateT s f)"
+  (=<<) :: forall a b. (a -> StateT s f b) -> StateT s f a -> StateT s f b
+  g =<< (StateT stateTFn)  = StateT (\s -> stateTFn s >>= mon)
+    where
+      mon :: Monad f => (a, s) -> f (b, s)
+      mon (a, s) = runStateT (g a) s
 
 -- | A `State'` is `StateT` specialised to the `ExactlyOne` functor.
-type State' s a =
-  StateT s ExactlyOne a
+type State' s a = StateT s ExactlyOne a
 
 -- | Provide a constructor for `State'` values
 --
 -- >>> runStateT (state' $ runState $ put 1) 0
 -- ExactlyOne  ((),1)
-state' ::
-  (s -> (a, s))
-  -> State' s a
-state' =
-  error "todo: Course.StateT#state'"
+state' :: (s -> (a, s)) -> State' s a
+state' f =  StateT (pure . f)
```

Comment:
nice!

Reply:
So.... if `state' f = StateT (pure . f)`, then `state' = StateT . pure`, yeah? :)

Reply:
oh don't look at this PR anymore, I changed most of this, need to finish it though